server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -1142,5 +1257,64 @@ synchronized void markItemAsDropped(int slot) {
             );
             itemResponses.add(BulkItemResponse.success(slot, docWriteRequest.opType(), dropped));
         }
+
+        /**
+         * Mark the document at the given slot in the bulk request as having failed in the ingest service. The document will be redirected
+         * to a data stream's failure store.
+         * @param slot the slot in the bulk request to redirect.
+         * @param targetIndexName the index that the document was targeting at the time of failure.
+         * @param e the failure encountered.
+         */
+        public void markItemForFailureStore(int slot, String targetIndexName, Exception e) {
```

Comment:
If you go with an separate listener, it would really cool if captured the `targetIndexName` on construction rather than requiring it be passed in here -- but I understand if that's not possible.

Reply:
This might be tricky, or I'm not fully understanding the suggestion. The target index name in this case is the index that we were targeting at the start of an ingest pipeline execution. Each recursive call to `executePipelines` has a chance that this value can change, mostly in the event of either a reroute processor or a processor updating the index target. We need to recapture the target index name on each top level pipeline execution.

I'm going to take a crack at the listener refactor though. If it's a bit much I'll file a followup issue/PR for it

Reply:
I attempted to unify all the composed logic in the service into one listener - it didn't go as cleanly as I had originally hoped. There are still a number of edge cases that will need to be solved and so I've shelved the refactor for now. Instead I'm going with a small change to lowers the repetition.